File: ts-plugins/src/eventstore/replay-service.ts
import type { VersionedTransitionEvent, ProjectionReducer } from './types.js';

/**
 * Replay service — reconstructs flow state at any version.
 *
 * Assumes each TRANSITION event stores a full snapshot of the state.
 * Returns the latest matching state at or before the requested version.
 *
 * If the event log is later changed to store diffs instead of full snapshots,
 * use {@link ProjectionReplayService} with a fold/reducer instead.
 */
export class ReplayService {
  stateAtVersion(events: readonly VersionedTransitionEvent[], flowId: string, targetVersion: number): string | null {
    const flowEvents = events
      .filter(e => e.flowId === flowId && e.type === 'TRANSITION' && e.version <= targetVersion)
      .sort((a, b) => a.version - b.version);
    if (flowEvents.length === 0) return null;
    return flowEvents[flowEvents.length - 1].to;
  }
}

/**
 * Projection replay service — fold/reducer model for custom aggregations.
 *
 * Unlike {@link ReplayService} which assumes full snapshots,
 * this service supports both full-snapshot and diff-based event logs.
 * `reducer.initialState()` returns the empty starting state,
 * `reducer.apply(state, event)` accumulates each event.
 *
 * Use for custom aggregations (transition count, cumulative metrics)
 * or when the event log stores diffs rather than full snapshots.
 */
export class ProjectionReplayService {
  stateAtVersion<T>(
    events: readonly VersionedTransitionEvent[],
    flowId: string,
    targetVersion: number,
    reducer: ProjectionReducer<T>,
  ): T {
    let state = reducer.initialState();
    for (const event of events) {
      if (event.flowId === flowId && event.version <= targetVersion) {
        state = reducer.apply(state, event);
      }
    }
    return state;
  }
}

File: lang/java-plugins/src/main/java/org/unlaxer/tramli/plugins/eventstore/ReplayService.java
package org.unlaxer.tramli.plugins.eventstore;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Replays versioned transition events for a flow.
 *
 * <p>Current semantics assume that each TRANSITION event stores a full snapshot
 * of the projected data at that version. Therefore {@link #stateAtVersion(List, String, long)}
 * returns the latest matching snapshot at or before the requested version.</p>
 *
 * <p>If the event log is later changed to store diffs instead of full snapshots,
 * this implementation must be replaced by a fold/reducer-based replay.</p>
 */
public final class ReplayService {
    public Map<String, Object> stateAtVersion(List<VersionedTransitionEvent> allEvents, String flowId, long version) {
        List<VersionedTransitionEvent> filtered = new ArrayList<>();
        for (VersionedTransitionEvent event : allEvents) {
            if (flowId.equals(event.flowId()) && event.version() <= version && "TRANSITION".equals(event.eventType())) {
                filtered.add(event);
            }
        }
        return filtered.isEmpty() ? Map.of() : filtered.get(filtered.size() - 1).dataSnapshot();
    }
}
